// @schibstedspain/ddd-react-redux/lib/mapUIServiceToProps.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _toArray2 = require('babel-runtime/helpers/toArray');

var _toArray3 = _interopRequireDefault(_toArray2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var toTitle = function toTitle(str) {
  var _str = (0, _toArray3.default)(str),
      first = _str[0],
      rest = _str.slice(1);

  return [first.toUpperCase()].concat((0, _toConsumableArray3.default)(rest)).join('');
};

var mapUIServiceToProps = function mapUIServiceToProps() {
  for (var _len = arguments.length, paths = Array(_len), _key = 0; _key < _len; _key++) {
    paths[_key] = arguments[_key];
  }

  return function (Target) {
    var _class, _temp;

    return _temp = _class = function (_Component) {
      (0, _inherits3.default)(Enhance, _Component);

      function Enhance(props, context) {
        (0, _classCallCheck3.default)(this, Enhance);

        var _this = (0, _possibleConstructorReturn3.default)(this, (Enhance.__proto__ || (0, _getPrototypeOf2.default)(Enhance)).call(this, props, context));

        _this.services = paths.reduce(function (props, path) {
          props['set' + toTitle(path) + 'UI'] = function (nextValue) {
            return context.store.dispatch({
              type: 'UI',
              payload: (0, _defineProperty3.default)({}, path, nextValue)
            });
          };
          return props;
        }, {});
        return _this;
      }

      (0, _createClass3.default)(Enhance, [{
        key: 'render',
        value: function render() {
          return _react2.default.createElement(Target, (0, _extends3.default)({}, this.services, this.props));
        }
      }]);
      return Enhance;
    }(_react.Component), _class.displayName = 'mapUIServiceToProps(' + Target.displayName + ')', _class.originalContextTypes = Target.originalContextTypes || Target.contextTypes, _class.contextTypes = {
      store: _propTypes2.default.object.isRequired
    }, _temp;
  };
};

exports.default = mapUIServiceToProps;
