// @s-ui/react-domain-connector/lib/Provider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _store = require('./store');

var _store2 = _interopRequireDefault(_store);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint no-console: 0 */
var Provider = function (_Component) {
  (0, _inherits3.default)(Provider, _Component);

  function Provider() {
    var _ref;

    (0, _classCallCheck3.default)(this, Provider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var _this = (0, _possibleConstructorReturn3.default)(this, (_ref = Provider.__proto__ || (0, _getPrototypeOf2.default)(Provider)).call.apply(_ref, [this].concat(args)));

    var _this$props = _this.props,
        i18n = _this$props.i18n,
        domain = _this$props.domain,
        store = _this$props.store;


    _this.i18n = i18n;
    _this.domain = domain;
    _this.store = store || (0, _store2.default)();
    return _this;
  }

  (0, _createClass3.default)(Provider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        i18n: this.i18n,
        domain: this.domain,
        store: this.store
      };
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var i18n = this.i18n,
          domain = this.domain,
          store = this.store;
      var nextI18n = nextProps.i18n,
          nextDomain = nextProps.domain,
          nextStore = nextProps.store;


      if (i18n !== nextI18n) {
        console.warn('<Provider> does not support changing `i18n` on the fly.');
      }

      if (domain !== nextDomain) {
        console.warn('<Provider> does not support changing `domain` on the fly.');
      }

      if (store !== nextStore) {
        console.warn('<Provider> does not support changing `store` on the fly.');
      }
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _reactRedux.Provider,
        { store: this.store },
        _react.Children.only(this.props.children)
      );
    }
  }]);
  return Provider;
}(_react.Component);

exports.default = Provider;


Provider.displayName = 'Provider';

Provider.propTypes = process.env.NODE_ENV !== "production" ? {
  i18n: _propTypes2.default.object.isRequired,
  children: _propTypes2.default.element.isRequired,
  domain: _propTypes2.default.object,
  store: _propTypes2.default.object
} : {};

Provider.defaultProps = {
  i18n: {
    t: function t(literal) {
      return literal;
    },
    n: function n(number) {
      return number;
    },
    c: function c(number) {
      return number;
    }
  },
  domain: {
    get: function get() {
      return _promise2.default.rejected('Provider default domain not implemented.');
    },
    fake: true
  }
};

Provider.childContextTypes = {
  i18n: _propTypes2.default.object,
  domain: _propTypes2.default.object,
  store: _propTypes2.default.object
};
