// @schibstedspain/ddd-react-redux/lib/store/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var actionsFactory = function actionsFactory(service) {
  return function (params) {
    return function (dispatch, getState, domain) {
      domain.get(service).execute(params).then(function (result) {
        dispatch({
          type: service,
          payload: { params: params, result: result, error: undefined }
        });
      }).catch(function (error) {
        return dispatch({
          type: service,
          payload: { params: params, result: undefined, error: error }
        });
      });
    };
  };
};

var actionsFromServices = function actionsFromServices(services) {
  return (0, _keys2.default)(services).reduce(function (acc, service) {
    acc[service] = actionsFactory(service);
    return acc;
  }, {});
};

var ui = function ui(path, nextValue) {
  return function (dispatch) {
    return dispatch({
      type: 'UI',
      payload: (0, _defineProperty3.default)({}, path, nextValue)
    });
  };
};

var actions = function actions(_ref) {
  var config = _ref.config,
      services = (0, _objectWithoutProperties3.default)(_ref, ['config']);
  return (0, _extends3.default)({}, actionsFromServices(services), { ui: ui });
};

exports.default = actions;
